### PR TITLE
Enable DisableDirectStreaming on a per request basis

### LINKED ISSUE #2166: Allow DisableDirectStreaming on a per request basis
Would allow ad-hoc capturing of requests and responses

### PR DESCRIPTION
**Pull request by russcam**
Closes #2166

### CODE REVIEW

File: src/Elasticsearch.Net/Configuration/Security/BasicAuthenticationCredentials.cs [hunk 1]
```diff
@@ -8,9 +8,6 @@ public class BasicAuthenticationCredentials
 
 		public string Password { get; set; }
 
-		public override string ToString()
-		{
-			return this.Username + ":" + this.Password;
-		}
+		public override string ToString() => this.Username + ":" + this.Password;
```
Comment (gmarz) [inline src/Elasticsearch.Net/Configuration/Security/BasicAuthenticationCredentials.cs]: nit: interpolate instead

File: src/Tests/ClientConcepts/LowLevel/DirectStreaming.cs [hunk 2]
```diff
@@ -79,5 +74,39 @@ public void EnableDirectStreamingOnSuccess()
...
+				r.ApiCall.RequestBodyInBytes.Should().BeNull();
+				r.ApiCall.ResponseBodyInBytes.Should().BeNull();
+			};
+
+			var client = TestClient.GetFixedReturnClient(new { }, 200, c => c.DisableDirectStreaming());
+			var response = client.Search<object>(s => s.RequestConfiguration(r => r.DisableDirectStreaming(false)));
+			assert(response);
+			response = client.SearchAsync<object>(s => s.RequestConfiguration(r => r.DisableDirectStreaming(false))).Result;
+			assert(response);
+		}
```
Comment (gmarz) [inline src/Tests/ClientConcepts/LowLevel/DirectStreaming.cs]: Can we add a test that shows `RequestConfiguration.DisableDirectStreaming` overrides `ConnectionSettings.DisableDirectStreaming`?

Comment (russcam) [inline src/Tests/ClientConcepts/LowLevel/DirectStreaming.cs]: These two tests check

1. when disable direct streaming is `false` globally, can be set to `true` locally
2. when disable direct streaming is `true` globally, can be set to `false` locally

Were you thinking of some other tests?

File: src/Elasticsearch.Net/Extensions/Extensions.cs [hunk 3]
```diff
@@ -1,32 +1,33 @@
...
 using System.Text;
-#if DOTNETCORE
-using System.Reflection;
-#endif
 
 namespace Elasticsearch.Net
 {
 	internal static class Extensions
 	{
+		private static readonly ConcurrentDictionary<Type, Dictionary<FieldInfo, string>> EnumStringValues =
```
Comment (Mpdreamz) [inline src/Elasticsearch.Net/Extensions/Extensions.cs]: These changes look good but should be part of a different commit.

File: src/Tests/ClientConcepts/LowLevel/DirectStreaming.cs (hunk 2 above)
Comment (gmarz) [inline src/Tests/ClientConcepts/LowLevel/DirectStreaming.cs]: My bad, I missed where you set DisableDirectStreaming globally in the second test.  LGTM 👍

File: src/Elasticsearch.Net/Extensions/Extensions.cs (hunk 3 above)
Comment (russcam) [inline src/Elasticsearch.Net/Extensions/Extensions.cs]: I'll pull them out and put them in a separate commit 👍

**Comment by russcam** [comment]
ported to `master` in 44c5c8499dcc5cc81edc3d4bbc990afa07c62383
